### PR TITLE
Transformer Prototype

### PR DESCRIPTION
**Pull request by georgimarian**
- add TransformerSeq2Seq model and Positional Encodings to /models folder
- add Transformer Test
- add dummy dataset, training_entry and vocabs
- add Transformer-based config
- use concept_identification_train flow with 2 extra flags for dummy dataset and transformer/lstm implementation

### CODE REVIEW

File: data_pipeline/dummy/dummy_training_entry.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from typing import List, Dict
+import penman
+from data_pipeline.concepts import get_ordered_concepts_ids
+from data_pipeline.adjacency_mat import get_adjacency_mat
+
+ROOT_ID = -1
```
Comment (Ranapop) [inline data_pipeline/dummy/dummy_training_entry.py]: why is the root id -1? where is it used?

File: data_pipeline/dummy/dummy_vocab.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+from typing import List, Tuple
+
+from data_pipeline.dummy.dummy_training_entry import DummyTrainingEntry
+from data_pipeline.vocab import read_cached_vocabs, cache_vocabs, build_vocab
+
+VOCAB_PATH = 'temp/vocabs'
+TOKENS_CACHE_FILE = 'letter_tokens.pickle'
+CONCEPTS_CACHE_FILE = 'letter_concepts.pickle'
+RELATIONS_CACHE_FILE = 'letter_relations.pickle'
```
Comment (Ranapop) [inline data_pipeline/dummy/dummy_vocab.py]: I really don't think the dummy vocabs should be cached, they are really small

File: model/transformer.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+               d_model: int,
+               dropout=0.1,
+               max_len=500):
+    super(PositionalEncoding, self).__init__()
+    self.dropout = nn.Dropout(p=dropout)
+
+    positional_encoding = torch.zeros(max_len, d_model)
+    position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+    div_term = torch.exp(torch.arange(
+        0, d_model, 2).float() * (-math.log(10000.0) / d_model))
```
Comment (Ranapop) [inline model/transformer.py]: what does 10000 represent?

File: model/transformer.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    self.pos_decoder = PositionalEncoding(embedding_dim, dropout)
+    # Vanilla Transformer
+    self.transformer = Transformer(embedding_dim, head_number, hidden_size,
+                                   dropout=dropout)
+    # Linear Transformation of Transformer Output
+    self.dense = nn.Linear(embedding_dim, output_vocab_size).to(device)
+    # Device to train on
+    self.device = device
+    # Masks
+    self.trg_mask = None
```
Comment (Ranapop) [inline model/transformer.py]: why do you need this variable?

File: model/transformer.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    if self.training:
+      gold_output_sequence = self.dec_embedding(gold_output_sequence)
+      gold_output_sequence = self.pos_decoder(gold_output_sequence)
+      self.trg_mask = self.make_triangular_mask(gold_output_sequence)
+      transformer_out = self.transformer(input_sequence,
+                                         gold_output_sequence,
+                                         tgt_mask=self.trg_mask)
+      logits = self.dense(transformer_out)
+      activated_outputs = torch.softmax(logits, dim=-1)
+      predictions = torch.argmax(activated_outputs, dim=-1)
```
Comment (Ranapop) [inline model/transformer.py]: this code could be extracted in a component; it's almost completely duplicated on the else branch (just send in the mask from the outside maybe)

**Review by Ranapop — CHANGES_REQUESTED** [review]

File: model/transformer.py (hunk 3 above)
Comment (georgimarian) [reply model/transformer.py]: it is used to compute the wavelength for the sine/cosine; taken from formula in _"Attention is All You Need"_
PE(pos, 2i) = sin(pos/10000^(2i/dmodel))
PE(pos, 2i+1) = cos(pos/10000^(2i/dmodel))

File: model/transformer.py (hunk 5 above)
Comment (georgimarian) [reply model/transformer.py]: extracted in _apply_layer_ function

File: data_pipeline/dummy/dummy_dataset.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            print("dummy sentence: ", sentence)
+            i = i + 1
+            concepts = [char for char in sentence[::-1]]
+            tokens = [char for char in sentence]
+            # Process the training entry (add EOS for sentence and concepts).
+            add_eos(tokens, concepts, EOS)
+            add_bos(tokens, BOS)
+            sentence, concepts = numericalize(tokens, concepts, vocabs)
+            print('sentence numericalized', sentence)
+            print('concepts numericalized', concepts)
```
Comment (Ranapop) [inline data_pipeline/dummy/dummy_dataset.py]: should you leave these prints here?

File: model/transformer.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+          output_sequence: input to the decoder
+          src_mask:
+          trg_mask: target mask for attention
+        Returns
+          logits: tensor of shape ...
+          predictions: tensor of shape ...
+        """
+        output_sequence = self.dec_embedding(output_sequence)
+        output_sequence = self.pos_decoder(output_sequence)
+        transformer_out = self.transformer_decoder(output_sequence,
```
Comment (Ranapop) [inline model/transformer.py:109]: where is the input seq embedded?

File: train_concept_identification.py [hunk 8]
```diff
@@ -303,6 +351,131 @@ def train_model(model: nn.Module,
...
+    if component:
+      # Select weights belonging to component
+      pretrained_dict = {k: v for k, v in pretrained_dict.items() if component in k}
+    model_dict = model.state_dict()
+    model_dict.update(pretrained_dict)
+    model.load_state_dict(model_dict)
+  return model
+
+
+def generate_sentences(len: int):
```
Comment (Ranapop) [inline train_concept_identification.py]: this should not be done in the train script -- rather should somehow be done in the dataset

File: train_concept_identification.py [hunk 9]
```diff
@@ -303,6 +351,131 @@ def train_model(model: nn.Module,
...
+
+
+def prepare_pretrain_model(criterion,
+                           scheduler,
+                           max_out_len: int,
+                           device: str,
+                           config: CfgNode,
+                           tensorboard_dir: str):
+    """
+      Prepare Transformer model for pretraining on Wikitext2 Dataset
```
Comment (Ranapop) [inline train_concept_identification.py]: what does this function actually do?

**Review by Ranapop — COMMENTED** [review]
The code looks overall good. The vocabs could probably be rewritten to extract the common code away, but this can also be done in another PR. 

Yet, the train script should definitely be simplified more -- the dummy data generation should not be there and we should discuss whether we will extract more code away or have different scripts for different training regimes.

Also, the model code could be better documented

File: model/transformer.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+class PositionalEncoding(nn.Module):
+    """
+    Positional encoding for transformer model
+    Code from https://pytorch.org/tutorials/beginner/transformer_tutorial.html
+    """
+
+    def __init__(self,
+                 d_model: int,
+                 dropout=0.1,
+                 max_len=500):
```
Comment (Ranapop) [inline model/transformer.py]: do you need a default value here? if yes, make it configurable, if not, delete it
Goes for:
* dropout
* max len

File: model/transformer.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        self.dropout = nn.Dropout(p=dropout)
+
+        positional_encoding = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+        div_term = torch.exp(torch.arange(
+            0, d_model, 2).float() * (-math.log(10000.0) / d_model))
+        positional_encoding[:, 0::2] = torch.sin(position * div_term)
+        positional_encoding[:, 1::2] = torch.cos(position * div_term)
+        positional_encoding = positional_encoding.unsqueeze(0).transpose(0, 1)
+        self.register_buffer('pe', positional_encoding)
```
Comment (Ranapop) [inline model/transformer.py:34]: if you took this code from a blog post or something it would be best to link it in a comment (maybe here, maybe at the top of the file)

File: train_concept_identification.py [hunk 12]
```diff
@@ -13,14 +15,23 @@
...
 from data_pipeline.glove_embeddings import GloVeEmbeddings
 from utils.extended_vocab_utils import construct_extended_vocabulary, numericalize_concepts
+from model.transformer import TransformerSeq2Seq
+from optimizer import NoamOpt
+from yacs.config import CfgNode
+from torchtext.datasets import WikiText2
+from torchtext.data.utils import get_tokenizer
+
 
 FLAGS = flags.FLAGS
```
Comment (Ranapop) [inline train_concept_identification.py:36]: There are quite a lot of flags now. I think we can move the flags and main function to a separate script. But this can be done in a different PR

**Review by Ranapop — CHANGES_REQUESTED** [review]
It looks overall good, I left a few coments.

The train script can be refactored in a different PR.

**Review by Ranapop — APPROVED** [review]

**Comment by Ranapop** [comment]
It looks good for now. We will have to refactor the datasets and vocabs, but it's not the highest priority at the moment.
